### DISCUSSION TITLE
How to handle mobile application lifecycle (Resumed, Paused, Destroyed...)?

### DISCUSSION THREAD

**Discussion by alessandrofrancesconi** [AvaloniaUI/Avalonia#14234] [Q&A]
I'm building a cross-platform Android / iOS app using Avalonia 11.
I need to know when the app goes in background (e.g. on Android, when the user clicks the "menu" button, or locks the device).

For Android, the application events are documented here: https://developer.android.com/guide/components/activities/activity-lifecycle
For iOS: https://developer.apple.com/documentation/uikit/app_and_environment/managing_your_app_s_life_cycle

How to handle these events for both the platforms, on Avalonia?

**Comment by maxkatz6** [comment]
You have access to the Android Activity and iOS UIApplicationDelegate - you can use the same APIs to handle these events and call your services/viewmodels to handle them.

**Reply by LyleRolleman-plst** [reply]
How can I hook into those events for iOS?

Android conveniently provides methods to override but the AppDelegate doesn't seem to provide the same. I'm interested in applicationWillTerminate

**Reply by maxkatz6** [reply]
You should be able to add this method to your AppDelegate with Export attribute:
```csharp
    public partial class AppDelegate : AvaloniaAppDelegate<App>
    {
        // protected override AppBuilder CustomizeAppBuilder {}

        [Export("applicationWillTerminate:")]
        public virtual void WillTerminate(UIApplication application)
        {
            
        }
    }
```

**Comment by maxkatz6** [comment]
Upd. In 11.1 we have some basic common events that can be handled for app activation/deactivation: https://docs.avaloniaui.net/docs/concepts/services/activatable-lifetime

**Reply by ah-pa** [reply]
Except that this is explicitly being disabled on iOS as far as I can see.

Passing null as the app delegate in UseiOS here

`

    public static class IOSApplicationExtensions
    {
        public static AppBuilder UseiOS(this AppBuilder builder, IAvaloniaAppDelegate appDelegate)
        {
            return builder
                .UseStandardRuntimePlatformSubsystem()
                .UseWindowingSubsystem(() => iOS.Platform.Register(appDelegate), "iOS")
                .UseSkia();
        }

        public static AppBuilder UseiOS(this AppBuilder builder) => UseiOS(builder, null!);
    }

`

means the lifetime event handlers are not registered so are never triggered.

Is there any particular reason for this?

Thanks

**Reply by maxkatz6** [reply]
Null is passed to keep old method in the public API, without breaking older apps.
But you are right - Avalonia itself should use new method, that's a bug. https://github.com/AvaloniaUI/Avalonia/pull/19550
